File: 4.1_find_if/find_if_Benchmark.cpp
#include <iostream>
#include <random>
#include <chrono>
#include "find_if_Benchmark.h"
#include "../utils/inputUtils.h"

using namespace std;

vector<int> numberOfIterations = vector<int>();

template<typename Iter, typename Pred>
Iter find_if(Iter begin, Iter beyond, Pred pred)
{
	int counter = 1;
	while(begin != beyond)
	{
		if (pred(*begin))
		{
			break;
		}

		++begin;
		++counter;
	}

	//We only measure it if we found the element
	if (begin != beyond)
	{
		numberOfIterations.push_back(counter);
	}

	return begin;
};

void find_if_Benchmark::run()
{
	cout << endl;
	cout << "--------- FIND_IF BENCHMARKING ---------" << endl;
	//Ask user for benchmark parameters
	cout << "How many runs do you want to perform? ";
	int numOfRuns = utils::inputUtils::getInt();

	cout << "How many numbers should the vector contain? ";
	int size = utils::inputUtils::getInt();

	cout << "Between 0 and which value should the random numbers be distributed? ";
	int range = utils::inputUtils::getInt();

	cout << "Start benchmark..." << endl;
	//Create vector with values
	vector<int> list = vector<int>();
	random_device rd;
	mt19937 eng(rd());
	uniform_int_distribution<> distr(0, range);

	for (int i = 0; i < size; i++)
	{
		list.push_back(distr(eng));
	}

	//Run benchmark
	vector<double> elapsedTimes = vector<double>();
	for (int i = 0; i < numOfRuns; i++)
	{
		int numberToFound = distr(eng);
		auto pred = [numberToFound] (const int &value) { return numberToFound == value; };
		auto start = chrono::high_resolution_clock::now();

		auto iter = find_if(list.begin(), list.end(), pred);

		auto finish = chrono::high_resolution_clock::now();
		auto elapsedTime = chrono::duration_cast<chrono::nanoseconds>(finish - start).count();

		//We only measure when we actually found the element
		if (iter != list.end())
		{
			elapsedTimes.push_back(elapsedTime);
		}
	}

	//Cacluate averages
	double avgElapsedTimeInNanoseconds = 0;
	double avgIterations = 0;
	for (unsigned int i = 0; i < elapsedTimes.size(); i++)
	{
		avgElapsedTimeInNanoseconds += elapsedTimes.at(i);
		avgIterations += numberOfIterations.at(i);
	}

	avgElapsedTimeInNanoseconds = avgElapsedTimeInNanoseconds / elapsedTimes.size();
	double avgElapsedTimeInMilliseconds = avgElapsedTimeInNanoseconds / 1000000;
	avgIterations = avgIterations / numberOfIterations.size();

	cout << "\nRESULTS" << endl;
	cout << "Average time: " << avgElapsedTimeInNanoseconds << "ns / " << avgElapsedTimeInMilliseconds << "ms" << endl;
	cout << "Average iterations: " << avgIterations << endl;
	double matching = elapsedTimes.size() / (double)numOfRuns * 100;
	cout << "Matching: " << matching << "% (" << elapsedTimes.size() << " of " << numOfRuns << ")" << endl;
}

File: 4.1_find_if/java/src/ch/advcpp/find_if_Benchmark.java
package ch.advcpp;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.concurrent.ThreadLocalRandom;

public class find_if_Benchmark<E>
{
    private List<Integer> numberOfIterations = new ArrayList<>();
    private List<Long> numberOfElapsedTimes = new ArrayList<>();
    private static final int NUMBER_OF_RUNS = 10000;
    private static final int RANGE = 1000000;

    private Iterator<E> find_if(Iterator<E> iter, Matcher<E> matcher)
    {
        int counter = 1;
        while(iter.hasNext())
        {
            if(matcher.match(iter.next()))
            {
                break;
            }

            counter++;
        }

        numberOfIterations.add(counter);
        return iter;
    }

    public void run()
    {
        //Create list with millions of random numbers between 0 and RANGE
        int size = RANGE;
        ArrayList<Integer> list = new ArrayList<>(size);
        for (int i = 0; i < size; i++)
        {
            list.add(ThreadLocalRandom.current().nextInt(0, RANGE));
        }

        for (int i = 0; i < NUMBER_OF_RUNS; i++)
        {
            int numberToFound = ThreadLocalRandom.current().nextInt(0, RANGE);
            Matcher<Integer> matcher = new EqualsMatcher<>(numberToFound);
            long start = System.nanoTime();
            Iterator<E> iter = find_if((Iterator<E>)list.iterator(), (Matcher<E>)matcher);
            long elapsedTime = (System.nanoTime() - start);
            numberOfElapsedTimes.add(elapsedTime);
        }

        //Calculate averages
        double avgElapsedTime = 0;
        double avgIterations = 0;
        if (numberOfElapsedTimes.size() != numberOfIterations.size())
        {
            System.out.println("Something went wrong..");
            return;
        }

        for (int i = 0; i < numberOfElapsedTimes.size(); i++)
        {
            avgElapsedTime += numberOfElapsedTimes.get(i);
            avgIterations += numberOfIterations.get(i);
        }

        avgElapsedTime = avgElapsedTime / numberOfElapsedTimes.size();
        avgIterations = avgIterations / numberOfIterations.size();

        System.out.println("Average time in nanoseconds: " + avgElapsedTime);
        System.out.println("Average iterations: " +  avgIterations);
    }
}
